client/client-lib/src/modules.rs

```diff
@@ -0,0 +1,260 @@
+use std::{collections::BTreeMap, io, sync::Arc};
+
+use impl_tools::autoimpl;
+
+use bitcoin::Amount;
+use thiserror::Error;
+
+/// Object-safe trait for things that can encode themselves
+///
+/// Like `rust-bitcoin`'s `consensus_encode`, but without generics,
+/// so can be used in `dyn` objects.
+pub trait Encodable {
+    fn consensus_encode(&self, writer: &mut dyn io::Write) -> Result<usize, std::io::Error>;
+}
+
+/// Something that can be an [`Input`] in a [`Transaction`]
+///
+/// General purpose code should use [`Input`] instead
+pub trait IInput: Encodable {}
+
+/// Something that can be an [`Output`] in a [`Transaction`]
+///
+/// General purpose code should use [`Output`] instead
+pub trait IOutput: Encodable {}
+
+/// An owned, immutable input to a [`Transaction`]
+#[autoimpl(Deref using self.0)]
+pub struct Input(Box<dyn IInput + Send + Sync + 'static>);
+
+/// An owned, immutable output of a [`Transaction`]
+#[autoimpl(Deref using self.0)]
+pub struct Output(Box<dyn IOutput + Send + Sync + 'static>);
+
+/// A spendable output - tracked and persisted by the client
+///
+/// Created by generating transaction [`Output`], spendable
+/// by converting to [`Input`].
+pub trait ISpendableOutput {
+    /// Module key
+    fn module_key(&self) -> &'static str;
+    fn amount(&self) -> Amount;
+    /// Prepare [`Input`] spending thish output in a transaction, and a key used to sign the [`Transaction`]
+    fn to_input(&self) -> (Input, bitcoin::KeyPair);
+
+    fn key(&self) -> String;
+}
+
+#[autoimpl(Deref using self.0)]
+pub struct SpendableOutput(Box<dyn ISpendableOutput + Sync + 'static>);
+
+#[derive(Debug, Error)]
+pub enum DecodeError {
+    #[error("boo")]
+    SomethingWentWrong,
+}
+
+/// Transaction, without a signature yet
+pub struct UnsignedTransaction {
+    inputs: Vec<SpendableOutput>,
```

Comment:
Note: while the tx is still being built (unsigned) the inputs are `SpendableOutput`s. Only when `UnsignedTransaction` is being signed and converted to `Transaction`, they are converted into `Input`s.